sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala

```diff
@@ -517,4 +517,57 @@ class SQLMetricsSuite extends SparkFunSuite with SQLMetricsTestUtils with Shared
   test("writing data out metrics with dynamic partition: parquet") {
     testMetricsDynamicPartition("parquet", "parquet", "t1")
   }
+
+  test("SPARK-25602: SparkPlan.getByteArrayRdd should not consume the input when not necessary") {
+    def checkFilterAndRangeMetrics(
+        df: DataFrame,
+        filterNumOutputs: Int,
+        rangeNumOutputs: Int): Unit = {
+      var filter: FilterExec = null
```

Comment:
what about something like this:
```
def collectExecNode[T](pf: PartialFunction[SparkPlan, T]): PartialFunction[SparkPlan, T] = {
        pf.orElse {
          case w: WholeStageCodegenExec =>
            w.child.collect(pf).head
        }
      }
      val range = df.queryExecution.executedPlan.collectFirst(
        collectExecNode { case r: RangeExec => r })
      val filter = df.queryExecution.executedPlan.collectFirst(
        collectExecNode { case f: FilterExec => f })
```

Reply:
In the future if we need to catch more nodes, we should abstract it. But for now it's only range and filter, I think it's ok.